### ISSUE TITLE
Access to /go/oauth/admin/clients ends in endless redirect behind SSL proxy

### ISSUE THREAD

**Issue by olvesh** [gocd/gocd#1418]
I have set up gocd using the official docker images behind a haproxy responsible for terminating the https traffic. 

When I access https://gocd.vimond.dev/go/oauth/admin/clients I am redirected to https://gocd.vimond.dev/go/oauth/admin/clients again and the browser warns me of a redirect loop.

The host gocd.vimond.dev is in my /etc/hosts file pointing to a vagrant instance.
The haproxuy https setup works as it should and I set the following options to add the X- headers:
```
  http-request add-header X-Proto https if { ssl_fc }
  http-request add-header X-Forwarded-Proto https if { ssl_fc }
  http-request add-header X-Forwarded-Ssl on if { ssl_fc }
  http-request set-header X-Forwarded-Port %[dst_port]
```
The siteurls are set, one with http://gocd.vimond.dev and the other with https://...

I also test direct access with curl to see if there is an error in the haproxy setup:
``` 
curl -v http://gocd.vimond.dev:32788/go/oauth/admin/clients \
 -H 'X-Forwarded-For: 192.168.33.60' \
 -H 'X-Forwarded-Proto: https' \
 -H 'X-Proto: https'
* Hostname was NOT found in DNS cache
*   Trying 192.168.33.60...
* Connected to gocd.vimond.dev (192.168.33.60) port 32788 (#0)
> GET /go/oauth/admin/clients HTTP/1.1
> User-Agent: curl/7.37.1
> Host: gocd.vimond.dev:32788
> Accept: */*
> X-Forwarded-For: 192.168.33.60
> X-Forwarded-Proto: https
> X-Proto: https
>
< HTTP/1.1 302 Found
< Date: Wed, 05 Aug 2015 14:53:21 GMT
< Set-Cookie: JSESSIONID=1ezhh94ezwho71vcsrk7izexlh;Path=/go;Expires=Wed, 19-Aug-2015 14:53:21 GMT
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< Location: https://gocd.vimond.dev/go/oauth/admin/clients
< Content-Length: 0
* Server Jetty(9.2.z-SNAPSHOT) is not blacklisted
< Server: Jetty(9.2.z-SNAPSHOT)
<
* Connection #0 to host gocd.vimond.dev left intact
```
(The port 32788 is the docker-assigned port mapping to 8153.)
As seen in the ```Location: https://gocd.vimond.dev/go/oauth/admin/clients``` response header I am redirected even if I say I am using https - and this location is fetched from the siteurls.

I hope I have done something wrong - but it smells more and more like a bug.

I also discussed the issue in #741  - and was encouraged by @ketan  to open a new issue for this case.

**Comment by ketan** [comment]
What are the values of `siteUrl` and `secureSiteUrl`. Could you share your server configuration. We need just the config without any `<agents/>` or `<pipelines/>`.

**Comment by olvesh** [comment]
```
<?xml version="1.0" encoding="utf-8"?>
<cruise xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="cruise-config.xsd"
        schemaVersion="75">
    <server artifactsdir="artifacts" 
            siteUrl="http://gocd.vimond.dev" 
            secureSiteUrl="https://gocd.vimond.dev"
            agentAutoRegisterKey="..." 
            commandRepositoryLocation="default"
            serverId="..."/>
</cruise>
```
No pipelines yet, as I am trying to evaluate gocd for our org. Oauth is important to us, so trying to set up that first..

**Comment by ketan** [comment]
Since you've set the secure site url to `https://gocd.vimond.dev`, go will redirect users to that URL, is HAProxy not setup to listen on port `443`, if HAProxy is listening on a different address/port, enter that address/port as the `siteUrl` and `secureSiteUrl`.

**Comment by ketan** [comment]
As you can see in the screenshot - I can see the page on build.go.cd just fine. I've already shared the apache configuration from that instance on the original issue.

The go configuration on build.go.cd is this -

```xml
<server siteUrl="https://build.go.cd" secureSiteUrl="https://build.go.cd" />
```

![screen shot 2015-08-06 at 12 28 19 pm](https://cloud.githubusercontent.com/assets/10598/9105575/a50948f0-3c36-11e5-82f1-94d3be0781f4.png)

**Comment by olvesh** [comment]
Haproxy listens to 443 just fine.
![gocd-https](https://cloud.githubusercontent.com/assets/220643/9105942/4f39466a-3c1c-11e5-9912-938a33a25677.png)

The apache configuration you shared connected to port  8154 - and if I haven't misunderstood that port expects https, while 8153 expects plain http. I have a haproxy setup with autodiscovery and I need to terminate ssl on haproxy and not in gocd.

**Comment by olvesh** [comment]
I managed to get haproxy to initiate a new https connection, but proxying https to http (8153) is obviously not supported.

If you close this issue then this should be noted in docs, as it is quite normal approach to terminate ssl and have unencrypted traffic internally.

Thanks for the help this far!

**Comment by ketan** [comment]
@olvesh - I think I was able to reproduce this problem at my end. While I get to the root cause of the issue, can you change your load balancer config to forward port 443 to go's 1854 -

This is the apache equivalent -

```apache
        ProxyPass         / https://localhost:8154/
        ProxyPassReverse  / https://localhost:8154/
        ProxyPreserveHost On
```

**Comment by olvesh** [comment]
Sure - thanks. That is what I tried to say I did in my previous post ;-)

**Comment by hany** [comment]
We are experiencing the same issue as well. Has there been any progress on this issue? We prefer to have HAProxy terminate all SSL connections.

**Comment by phylake** [comment]
I'm experiencing something similar. Login redirects to http when both `siteUrl` and `secureSiteUrl` are https

**Comment by gregorygtseng** [comment]
Looks like this was fixed in the 16.1.0-2855 release.
